sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala

```diff
@@ -409,8 +411,26 @@ private[sql] class DataFrameImpl protected[sql](
   override def toJSON: RDD[String] = {
     val rowSchema = this.schema
     this.mapPartitions { iter =>
-      val jsonFactory = new JsonFactory()
-      iter.map(JsonRDD.rowToJSON(rowSchema, jsonFactory))
+      val writer = new CharArrayWriter()
+      // create the Generator without separator inserted between 2 records
+      val gen = new JsonFactory().createGenerator(writer).setRootValueSeparator(null)
```

Comment:
A ` `(single space) will be inserted between 2 json records (in string) by default; 
Previously, we create the `generator` object for each of record in serialization, hence it's OK with default object separator.
However, in this PR, we are trying to using a single `generator` for all of the records, that's why we need to set it as null, which means nothing will be inserted between records.